From 7094bd6d786106f991a2145e48092a59821e8c01 Mon Sep 17 00:00:00 2001
From: Omkar Chandorkar <gotenksIN@aospa.co>
Date: Thu, 27 Aug 2026 13:38:09 +0530
Subject: [PATCH 2/2] SystemUI: Modernize Wi-Fi generation badge for Android 17
 icons

Update Wi-Fi standard badge icons to match the Android 17 status bar
geometry and integrate with the modern status bar pipeline.

- Regenerate Wi-Fi 4, 5, 6, and 7 badges on the 16.21x12.0 canvas,
  positioned on the lower-right of the icon.
- Regenerate ic_statusbar_wifi_standard with Android 17 style Wi-Fi icon.
- Remove WifiStandardImageView and direct view callbacks/observers.
- Integrate wifiStandardIcon flow into WifiViewModel, WifiViewBinder,
  ModernStatusBarWifiView, and Compose WifiIcon.
- Hide Wi-Fi generation badge when no internet / exclamation is shown.

Change-Id: Ia3406ed890d98ec5a7d18edd82c7b60b89fe95af
Signed-off-by: Omkar Chandorkar <gotenksIN@aospa.co>
---
 .../wifi/ui/viewmodel/WifiViewModelTest.kt    |   2 +
 .../drawable/ic_statusbar_wifi_standard.xml   |  51 +++---
 .../res/drawable/ic_wifi_standard_4.xml       |  12 +-
 .../res/drawable/ic_wifi_standard_5.xml       |  12 +-
 .../res/drawable/ic_wifi_standard_6.xml       |  12 +-
 .../res/drawable/ic_wifi_standard_7.xml       |  11 +-
 .../res/layout/new_status_bar_wifi_group.xml  |   2 +
 .../layout/status_bar_wifi_group_inner.xml    |  10 +-
 .../statusbar/WifiStandardImageView.kt        | 163 ------------------
 .../repository/prod/WifiRepositoryImpl.kt     |   1 +
 .../wifi/shared/model/WifiNetworkModel.kt     |  16 +-
 .../pipeline/wifi/ui/binder/WifiViewBinder.kt |   8 +
 .../pipeline/wifi/ui/compose/WifiIcon.kt      |  25 ++-
 .../pipeline/wifi/ui/model/WifiIcon.kt        |   9 +-
 .../wifi/ui/view/ModernStatusBarWifiView.kt   |   7 +
 .../wifi/ui/viewmodel/WifiViewModel.kt        |  48 ++++++
 .../wifi/ui/viewmodel/WifiViewModelCommon.kt  |   4 +
 .../ui/viewmodel/SystemStatusIconViewModel.kt |   1 +
 .../wifi/ui/viewmodel/WifiIconViewModel.kt    |   6 +
 .../ui/view/ModernStatusBarWifiViewTest.kt    |   6 +-
 .../WifiViewModelIconParameterizedTest.kt     |   6 +-
 .../domain/interactor/WifiInteractorKosmos.kt |   3 +
 .../wifi/ui/viewmodel/WifiViewModelKosmos.kt  |   2 +
 23 files changed, 191 insertions(+), 226 deletions(-)
 delete mode 100644 packages/SystemUI/src/com/android/systemui/statusbar/WifiStandardImageView.kt

diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelTest.kt b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelTest.kt
index a267b92b0..6315a807c 100644
--- a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelTest.kt
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelTest.kt
@@ -24,6 +24,7 @@ import com.android.systemui.common.shared.model.ContentDescription.Companion.loa
 import com.android.systemui.coroutines.collectLastValue
 import com.android.systemui.kosmos.testScope
 import com.android.systemui.log.table.logcatTableLogBuffer
+import com.android.systemui.shared.settings.data.repository.fakeSecureSettingsRepository
 import com.android.systemui.statusbar.connectivity.WifiIcons
 import com.android.systemui.statusbar.phone.StatusBarLocation
 import com.android.systemui.statusbar.pipeline.airplane.ui.viewmodel.AirplaneModeViewModel
@@ -389,6 +390,7 @@ class WifiViewModelTest : SysuiTestCase() {
                 interactor,
                 testScope.backgroundScope,
                 wifiConstants,
+                kosmos.fakeSecureSettingsRepository,
             )
     }
 
diff --git a/packages/SystemUI/res/drawable/ic_statusbar_wifi_standard.xml b/packages/SystemUI/res/drawable/ic_statusbar_wifi_standard.xml
index 0944e0206..920c30837 100644
--- a/packages/SystemUI/res/drawable/ic_statusbar_wifi_standard.xml
+++ b/packages/SystemUI/res/drawable/ic_statusbar_wifi_standard.xml
@@ -1,42 +1,35 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
-    Copyright (c) 2019, The Linux Foundation. All rights reserved.
+    Copyright (C) 2026 Paranoid Android
 
-    Redistribution and use in source and binary forms, with or without
-    modification, are permitted provided that the following conditions are
-    met:
-        * Redistributions of source code must retain the above copyright
-          notice, this list of conditions and the following disclaimer.
-        * Redistributions in binary form must reproduce the above
-          copyright notice, this list of conditions and the following
-          disclaimer in the documentation and/or other materials provided
-          with the distribution.
-        * Neither the name of The Linux Foundation nor the names of its
-          contributors may be used to endorse or promote products derived
-          from this software without specific prior written permission.
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
 
-    THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
-    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
-    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
-    ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
-    BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
-    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
-    BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
-    WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
-    OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
-    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
 -->
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-    android:width="26dp"
+    android:width="24dp"
     android:height="24dp"
-    android:viewportWidth="26"
-    android:viewportHeight="24"
+    android:viewportWidth="24.0"
+    android:viewportHeight="24.0"
     android:tint="?android:attr/colorControlNormal">
     <path
         android:fillColor="#FFFFFFFF"
-        android:pathData="m 25.626571,16.969114 v 0 q 0,1.36304 -0.84885,2.1693 -0.84886,0.80626 -2.32751,0.80626 v 0 q -1.29001,0 -2.06585,-0.58112 -0.77583,-0.58112 -0.95838,-1.6825 v 0 l 1.70988,-0.13995 q 0.13387,0.54765 0.47462,0.79713 0.34076,0.24949 0.85799,0.24949 v 0 q 0.63892,0 1.01923,-0.4077 0.38031,-0.40769 0.38031,-1.1744 v 0 q 0,-0.67543 -0.35901,-1.08008 -0.35902,-0.40465 -1.00402,-0.40465 v 0 q -0.71195,0 -1.16223,0.55373 v 0 h -1.66729 l 0.29817,-4.825391 h 5.15397 v 1.27176 h -3.6023 l -0.13996,2.166251 q 0.62067,-0.54764 1.55167,-0.54764 v 0 q 1.22308,0 1.95632,0.76062 0.73324,0.76062 0.73324,2.06889 z"/>
+        android:pathData="M17.92,13.89C18.62,13.18 18.56,12.01 17.69,11.53C16.00,10.60 14.06,10.06 12.01,10.06C9.95,10.06 8.00,10.60 6.32,11.53C5.45,12.01 5.39,13.18 6.09,13.89C6.66,14.46 7.57,14.52 8.30,14.16C9.42,13.61 10.68,13.30 12.01,13.30C13.33,13.30 14.59,13.61 15.71,14.16C16.44,14.52 17.35,14.46 17.92,13.89Z" />
+    <path
+        android:fillColor="#FFFFFFFF"
+        android:pathData="M22.07,9.74C22.73,9.07 22.71,7.98 21.95,7.44C19.16,5.41 15.72,4.20 12.01,4.20C8.29,4.20 4.86,5.40 2.06,7.44C1.31,7.98 1.28,9.07 1.94,9.74C2.56,10.35 3.53,10.38 4.23,9.88C6.44,8.35 9.12,7.45 12.01,7.45C14.89,7.45 17.57,8.35 19.78,9.88C20.48,10.38 21.46,10.35 22.07,9.74Z" />
+    <path
+        android:fillColor="#FFFFFFFF"
+        android:pathData="M12.01,19.80C13.08,19.80 13.96,18.93 13.96,17.85C13.96,16.77 13.08,15.90 12.01,15.90C10.93,15.90 10.06,16.77 10.06,17.85C10.06,18.93 10.93,19.80 12.01,19.80Z" />
     <path
         android:fillColor="#FFFFFFFF"
-        android:pathData="m 12.998641,2.099185 c -7.3000003,0 -12.0991851,4.002174 -12.59918509,4.402174 L 12.998641,21.998642 16.944293,17.148098 a 7.7288137,8.3389834 0 0 1 -0.06114,-0.978261 7.7288137,8.3389834 0 0 1 7.634511,-8.335597 l 1.084238,-1.332881 c -0.499999,-0.4 -5.303261,-4.402174 -12.603261,-4.402174 z"/>
+        android:pathData="M24.00,20.01L24.00,20.01Q24.00,20.95 23.43,21.47Q22.86,22.00 21.84,22.00L21.84,22.00Q20.72,22.00 20.11,21.27Q19.52,20.56 19.52,19.15L19.52,19.15Q19.52,17.61 20.13,16.81Q20.74,16.03 21.88,16.03L21.88,16.03Q22.69,16.03 23.15,16.36Q23.64,16.68 23.83,17.37L23.83,17.37L22.62,17.52Q22.45,16.94 21.86,16.94L21.86,16.94Q21.34,16.94 21.05,17.41Q20.75,17.88 20.75,18.81L20.75,18.81Q20.96,18.52 21.32,18.35Q21.69,18.18 22.14,18.18L22.14,18.18Q23.00,18.18 23.49,18.67Q24.00,19.17 24.00,20.01ZM22.71,20.05L22.71,20.05Q22.71,19.56 22.47,19.30Q22.22,19.04 21.78,19.04L21.78,19.04Q21.35,19.04 21.10,19.28Q20.84,19.53 20.84,19.93L20.84,19.93Q20.84,20.43 21.11,20.76Q21.37,21.09 21.80,21.09L21.80,21.09Q22.23,21.09 22.48,20.82Q22.71,20.54 22.71,20.05Z" />
 </vector>
diff --git a/packages/SystemUI/res/drawable/ic_wifi_standard_4.xml b/packages/SystemUI/res/drawable/ic_wifi_standard_4.xml
index d4c0d83dd..b5ebd751d 100644
--- a/packages/SystemUI/res/drawable/ic_wifi_standard_4.xml
+++ b/packages/SystemUI/res/drawable/ic_wifi_standard_4.xml
@@ -1,16 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
      Copyright (C) 2022 Project Kaleidoscope
-               (C) 2022 Paranoid Android
+               (C) 2022-2026 Paranoid Android
 
      SPDX-License-Identifier: Apache-2.0
 -->
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-    android:width="24dp"
-    android:height="24dp"
-    android:viewportWidth="24"
-    android:viewportHeight="24">
+    android:width="17.0dp"
+    android:height="12.58dp"
+    android:viewportWidth="16.21"
+    android:viewportHeight="12.0">
     <path
         android:fillColor="@android:color/white"
-        android:pathData="M 17.37 19.6795 L 17.37 18.85 L 21.318 13.1905 L 22.368 13.1905 L 22.368 18.661 L 23.5335 18.661 L 23.5335 19.711 L 22.389 19.711 L 22.389 21.5905 L 21.15 21.5905 L 21.15 19.6795 Z M 18.7245 18.6295 L 21.15 18.6295 L 21.15 15.2695 Z"/>
+        android:pathData="M 12.383,10.817 L 12.383,10.303 L 14.827,6.800 L 15.477,6.800 L 15.477,10.186 L 16.200,10.186 L 16.200,10.836 L 15.491,10.836 L 15.491,12.000 L 14.724,12.000 L 14.724,10.817 Z M 13.222,10.167 L 14.724,10.167 L 14.724,8.087 Z" />
 </vector>
diff --git a/packages/SystemUI/res/drawable/ic_wifi_standard_5.xml b/packages/SystemUI/res/drawable/ic_wifi_standard_5.xml
index ce06aa46f..a17192c15 100644
--- a/packages/SystemUI/res/drawable/ic_wifi_standard_5.xml
+++ b/packages/SystemUI/res/drawable/ic_wifi_standard_5.xml
@@ -1,16 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
      Copyright (C) 2022 Project Kaleidoscope
-               (C) 2022 Paranoid Android
+               (C) 2022-2026 Paranoid Android
 
      SPDX-License-Identifier: Apache-2.0
 -->
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-    android:width="24dp"
-    android:height="24dp"
-    android:viewportWidth="24"
-    android:viewportHeight="24">
+    android:width="17.0dp"
+    android:height="12.58dp"
+    android:viewportWidth="16.21"
+    android:viewportHeight="12.0">
     <path
         android:fillColor="@android:color/white"
-        android:pathData="M 23.7861 19.4085 L 23.7861 19.4085 Q 23.7861 20.5739 23.0602 21.263 Q 22.3343 21.953 21.0697 21.953 L 21.0697 21.953 Q 19.9668 21.953 19.3033 21.4562 Q 18.6407 20.9586 18.4842 20.0173 L 18.4842 20.0173 L 19.9463 19.8976 Q 20.0608 20.3661 20.3515 20.579 Q 20.6431 20.7927 21.0851 20.7927 L 21.0851 20.7927 Q 21.6315 20.7927 21.9572 20.4439 Q 22.2821 20.0951 22.2821 19.4401 L 22.2821 19.4401 Q 22.2821 18.8621 21.9752 18.5167 Q 21.6682 18.1705 21.1168 18.1705 L 21.1168 18.1705 Q 20.508 18.1705 20.1233 18.6441 L 20.1233 18.6441 L 18.6971 18.6441 L 18.9528 14.5179 L 23.3594 14.5179 L 23.3594 15.6055 L 20.2789 15.6055 L 20.1592 17.4574 Q 20.6901 16.9897 21.4861 16.9897 L 21.4861 16.9897 Q 22.5318 16.9897 23.1585 17.6395 Q 23.7861 18.2902 23.7861 19.4085 Z"/>
+        android:pathData="M 16.200,10.221 L 16.200,10.221 Q 16.200,11.036 15.694,11.518 Q 15.186,12.000 14.301,12.000 L 14.301,12.000 Q 13.530,12.000 13.065,11.652 Q 12.603,11.305 12.493,10.647 L 12.493,10.647 L 13.515,10.563 Q 13.595,10.890 13.799,11.039 Q 14.003,11.188 14.312,11.188 L 14.312,11.188 Q 14.693,11.188 14.921,10.945 Q 15.149,10.700 15.149,10.242 L 15.149,10.242 Q 15.149,9.838 14.933,9.596 Q 14.719,9.354 14.334,9.354 L 14.334,9.354 Q 13.909,9.354 13.640,9.686 L 13.640,9.686 L 12.641,9.686 L 12.820,6.800 L 15.902,6.800 L 15.902,7.560 L 13.747,7.560 L 13.664,8.856 Q 14.035,8.529 14.591,8.529 L 14.591,8.529 Q 15.323,8.529 15.762,8.983 Q 16.200,9.438 16.200,10.221 Z" />
 </vector>
diff --git a/packages/SystemUI/res/drawable/ic_wifi_standard_6.xml b/packages/SystemUI/res/drawable/ic_wifi_standard_6.xml
index 2e51071d9..b6e592dc9 100644
--- a/packages/SystemUI/res/drawable/ic_wifi_standard_6.xml
+++ b/packages/SystemUI/res/drawable/ic_wifi_standard_6.xml
@@ -1,16 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
      Copyright (C) 2022 Project Kaleidoscope
-               (C) 2022 Paranoid Android
+               (C) 2022-2026 Paranoid Android
 
      SPDX-License-Identifier: Apache-2.0
 -->
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-    android:width="24dp"
-    android:height="24dp"
-    android:viewportWidth="24"
-    android:viewportHeight="24">
+    android:width="17.0dp"
+    android:height="12.58dp"
+    android:viewportWidth="16.21"
+    android:viewportHeight="12.0">
     <path
         android:fillColor="@android:color/white"
-        android:pathData="M 23.8067 19.3448 L 23.8067 19.3448 Q 23.8067 20.5343 23.0721 21.2102 Q 22.3365 21.8872 21.0423 21.8872 L 21.0423 21.8872 Q 19.5903 21.8872 18.8114 20.9647 Q 18.0325 20.0425 18.0325 18.2302 L 18.0325 18.2302 Q 18.0325 16.2374 18.8231 15.2302 Q 19.6137 14.224 21.083 14.224 L 21.083 14.224 Q 22.1271 14.224 22.731 14.6418 Q 23.3338 15.0589 23.5847 15.936 L 23.5847 15.936 L 22.0396 16.1319 Q 21.8176 15.3972 21.0478 15.3972 L 21.0478 15.3972 Q 20.3889 15.3972 20.0126 15.9947 Q 19.6362 16.5912 19.6362 17.8069 L 19.6362 17.8069 Q 19.8989 17.4107 20.3656 17.1995 Q 20.8321 16.9875 21.4214 16.9875 L 21.4214 16.9875 Q 22.5234 16.9875 23.165 17.622 Q 23.8067 18.2564 23.8067 19.3448 Z M 22.1614 19.3872 L 22.1614 19.3872 Q 22.1614 18.7527 21.8383 18.417 Q 21.5144 18.0821 20.9485 18.0821 L 20.9485 18.0821 Q 20.4061 18.0821 20.0803 18.3962 Q 19.7536 18.7103 19.7536 19.2283 L 19.7536 19.2283 Q 19.7536 19.8781 20.0948 20.3041 Q 20.4359 20.7292 20.99 20.7292 L 20.99 20.7292 Q 21.5433 20.7292 21.8529 20.3728 Q 22.1614 20.0162 22.1614 19.3872 Z"/>
+        android:pathData="M 16.200,10.274 L 16.200,10.274 Q 16.200,11.082 15.701,11.541 Q 15.203,12.000 14.324,12.000 L 14.324,12.000 Q 13.340,12.000 12.810,11.375 Q 12.282,10.748 12.282,9.519 L 12.282,9.519 Q 12.282,8.166 12.819,7.483 Q 13.354,6.800 14.352,6.800 L 14.352,6.800 Q 15.060,6.800 15.470,7.083 Q 15.880,7.366 16.050,7.961 L 16.050,7.961 L 15.001,8.095 Q 14.850,7.597 14.328,7.597 L 14.328,7.597 Q 13.881,7.597 13.626,8.001 Q 13.371,8.406 13.371,9.231 L 13.371,9.231 Q 13.548,8.963 13.865,8.820 Q 14.182,8.675 14.582,8.675 L 14.582,8.675 Q 15.330,8.675 15.765,9.106 Q 16.200,9.536 16.200,10.274 Z M 15.083,10.304 L 15.083,10.304 Q 15.083,9.873 14.865,9.646 Q 14.645,9.418 14.260,9.418 L 14.260,9.418 Q 13.893,9.418 13.672,9.631 Q 13.450,9.845 13.450,10.196 L 13.450,10.196 Q 13.450,10.637 13.682,10.926 Q 13.912,11.214 14.288,11.214 L 14.288,11.214 Q 14.664,11.214 14.874,10.972 Q 15.083,10.730 15.083,10.304 Z" />
 </vector>
diff --git a/packages/SystemUI/res/drawable/ic_wifi_standard_7.xml b/packages/SystemUI/res/drawable/ic_wifi_standard_7.xml
index 9450d7ac0..5c53fb48b 100644
--- a/packages/SystemUI/res/drawable/ic_wifi_standard_7.xml
+++ b/packages/SystemUI/res/drawable/ic_wifi_standard_7.xml
@@ -5,12 +5,11 @@
      SPDX-License-Identifier: Apache-2.0
 -->
 <vector xmlns:android="http://schemas.android.com/apk/res/android"
-    android:width="24dp"
-    android:height="24dp"
-    android:viewportWidth="24"
-    android:viewportHeight="24">
+    android:width="17.0dp"
+    android:height="12.58dp"
+    android:viewportWidth="16.21"
+    android:viewportHeight="12.0">
     <path
-        android:pathData="M 23.843 16.451 L 20.411 22 L 18.267 21.976 L 21.831 16.377 L 18 16.335 L 18.014 15 L 23.858 15.066 Z"
         android:fillColor="@android:color/white"
-        android:strokeWidth="1"/>
+        android:pathData="M 16.189,7.877 L 13.640,12.000 L 12.046,11.982 L 14.695,7.822 L 11.849,7.792 L 11.859,6.800 L 16.200,6.848 Z" />
 </vector>
diff --git a/packages/SystemUI/res/layout/new_status_bar_wifi_group.xml b/packages/SystemUI/res/layout/new_status_bar_wifi_group.xml
index 753ba2f21..eaa211e2f 100644
--- a/packages/SystemUI/res/layout/new_status_bar_wifi_group.xml
+++ b/packages/SystemUI/res/layout/new_status_bar_wifi_group.xml
@@ -21,6 +21,8 @@
     android:id="@+id/wifi_combo"
     android:layout_width="wrap_content"
     android:layout_height="match_parent"
+    android:clipChildren="false"
+    android:clipToPadding="false"
     android:gravity="center_vertical" >
 
     <include layout="@layout/status_bar_wifi_group_inner" />
diff --git a/packages/SystemUI/res/layout/status_bar_wifi_group_inner.xml b/packages/SystemUI/res/layout/status_bar_wifi_group_inner.xml
index 552ee5051..f8c1cc2e5 100644
--- a/packages/SystemUI/res/layout/status_bar_wifi_group_inner.xml
+++ b/packages/SystemUI/res/layout/status_bar_wifi_group_inner.xml
@@ -24,6 +24,8 @@
         android:layout_width="wrap_content"
         android:layout_height="match_parent"
         android:gravity="center_vertical"
+        android:clipChildren="false"
+        android:clipToPadding="false"
         android:layout_marginStart="2.5sp"
     >
         <FrameLayout
@@ -52,15 +54,19 @@
             android:id="@+id/wifi_combo"
             android:layout_height="wrap_content"
             android:layout_width="wrap_content"
+            android:clipChildren="false"
+            android:clipToPadding="false"
             android:gravity="center_vertical" >
             <com.android.systemui.statusbar.AlphaOptimizedImageView
                 android:id="@+id/wifi_signal"
                 android:layout_height="@dimen/status_bar_wifi_signal_size"
                 android:layout_width="@dimen/status_bar_wifi_signal_size" />
-            <com.android.systemui.statusbar.WifiStandardImageView
+            <com.android.systemui.statusbar.AlphaOptimizedImageView
                 android:id="@+id/wifi_standard"
                 android:layout_height="@dimen/status_bar_wifi_signal_size"
-                android:layout_width="@dimen/status_bar_wifi_signal_size" />
+                android:layout_width="@dimen/status_bar_wifi_signal_size"
+                android:translationY="1.5dp"
+                android:visibility="gone" />
         </FrameLayout>
 
         <View
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/WifiStandardImageView.kt b/packages/SystemUI/src/com/android/systemui/statusbar/WifiStandardImageView.kt
deleted file mode 100644
index 172941be3..000000000
--- a/packages/SystemUI/src/com/android/systemui/statusbar/WifiStandardImageView.kt
+++ /dev/null
@@ -1,163 +0,0 @@
-/*
- * Copyright (C) 2023 The risingOS Android Project
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License
- */
-package com.android.systemui.statusbar
-
-import android.content.Context
-import android.database.ContentObserver
-import android.os.UserHandle
-import android.net.ConnectivityManager
-import android.net.Network
-import android.net.NetworkCapabilities
-import android.net.NetworkRequest
-import android.net.wifi.WifiManager
-import android.provider.Settings
-import android.util.AttributeSet
-import android.widget.ImageView
-import com.android.systemui.res.R
-
-const val TUNER_KEY = "wifi_standard"
-
-class WifiStandardImageView @JvmOverloads constructor(
-    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
-) : ImageView(context, attrs, defStyleAttr) {
-
-    private val connectivityManager: ConnectivityManager by lazy {
-        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
-    }
-    private val wifiManager: WifiManager by lazy {
-        context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
-    }
-    private var networkCallback: ConnectivityManager.NetworkCallback? = null
-    private var wifiStandardEnabled = false
-    private var isRegistered = false
-
-    init {
-        setupContentObserver()
-    }
-
-    override fun onAttachedToWindow() {
-        super.onAttachedToWindow()
-        showWifiStandard()
-    }
-
-    override fun onDetachedFromWindow() {
-        super.onDetachedFromWindow()
-        unregisterNetworkCallback()
-    }
-
-    private fun setupContentObserver() {
-        val showWifiStandardIcon = Settings.Secure.getUriFor(TUNER_KEY)
-        val contentObserver = object: ContentObserver(null) {
-            override fun onChange(selfChange: Boolean) {
-                wifiStandardEnabled = Settings.Secure.getIntForUser(context.contentResolver,
-                        TUNER_KEY, 0, UserHandle.USER_CURRENT) == 1
-                if (wifiStandardEnabled) {
-                    showWifiStandard()
-                } else {
-                    unregisterNetworkCallback()
-                }
-            }
-        }
-        context.contentResolver.registerContentObserver(
-                showWifiStandardIcon, false, contentObserver, UserHandle.USER_CURRENT)
-        contentObserver.onChange(true)
-    }
-
-    private fun showWifiStandard() {
-        if (!wifiStandardEnabled || networkCallback != null) return
-        networkCallback = object : ConnectivityManager.NetworkCallback() {
-            override fun onAvailable(network: Network) {
-                updateWifiStandard(network)
-            }
-
-            override fun onUnavailable() {
-                updateWifiStandard(null)
-            }
-
-            override fun onCapabilitiesChanged(
-                network: Network,
-                networkCapabilities: NetworkCapabilities
-            ) {
-                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
-                    updateWifiStandard(network)
-                }
-            }
-        }
-        registerNetworkCallback()
-    }
-
-    private fun updateWifiStandard(network: Network?) {
-        val wifiStandard = if (network != null) getWifiStandard(network) else -1
-        updateIcon(wifiStandard)
-    }
-
-    private fun getWifiStandard(network: Network): Int {
-        val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
-        return if (networkCapabilities != null &&
-                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) &&
-                networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)) {
-            val wifiInfo = wifiManager.connectionInfo
-            wifiInfo.wifiStandard
-        } else {
-            -1
-        }
-    }
-
-    private fun updateIcon(wifiStandard: Int) {
-        val drawableId = getDrawableForWifiStandard(wifiStandard)
-        if (!wifiStandardEnabled || drawableId == 0) {
-            post { visibility = GONE }
-            return
-        }
-        post {
-            setImageResource(drawableId)
-            visibility = VISIBLE
-        }
-    }
-
-    private fun getDrawableForWifiStandard(wifiStandard: Int): Int {
-        return when (wifiStandard) {
-            4 -> R.drawable.ic_wifi_standard_4
-            5 -> R.drawable.ic_wifi_standard_5
-            6 -> R.drawable.ic_wifi_standard_6
-            7 -> R.drawable.ic_wifi_standard_7
-            else -> 0
-        }
-    }
-
-    private fun registerNetworkCallback() {
-        if (isRegistered || networkCallback == null) return
-        val networkRequest = NetworkRequest.Builder()
-            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
-            .build()
-        connectivityManager.registerNetworkCallback(networkRequest, networkCallback!!)
-        isRegistered = true
-    }
-
-    private fun unregisterNetworkCallback() {
-        if (!isRegistered || networkCallback == null) return
-        try {
-            networkCallback?.let { connectivityManager.unregisterNetworkCallback(it) }
-            post {
-                visibility = GONE
-            }
-        } catch (e: IllegalArgumentException) {
-        } finally {
-            networkCallback = null
-            isRegistered = false
-        }
-    }
-}
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/data/repository/prod/WifiRepositoryImpl.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/data/repository/prod/WifiRepositoryImpl.kt
index ddfc4ac70..631b64048 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/data/repository/prod/WifiRepositoryImpl.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/data/repository/prod/WifiRepositoryImpl.kt
@@ -415,6 +415,7 @@ constructor(
             level = this.level,
             ssid = this.title,
             hotspotDeviceType = hotspotDeviceType,
+            wifiStandard = this.getConnectedInfo()?.wifiStandard ?: ScanResult.WIFI_STANDARD_UNKNOWN,
         )
     }
 
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/shared/model/WifiNetworkModel.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/shared/model/WifiNetworkModel.kt
index 3416d43ea..1283623ec 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/shared/model/WifiNetworkModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/shared/model/WifiNetworkModel.kt
@@ -16,6 +16,7 @@
 
 package com.android.systemui.statusbar.pipeline.wifi.shared.model
 
+import android.net.wifi.ScanResult
 import android.net.wifi.WifiManager
 import android.net.wifi.WifiManager.UNKNOWN_SSID
 import android.net.wifi.sharedconnectivity.app.NetworkProviderInfo
@@ -243,6 +244,13 @@ sealed class WifiNetworkModel : Diffable<WifiNetworkModel> {
          * isn't a hotspot connection.
          */
         val hotspotDeviceType: HotspotDeviceType,
+
+        /**
+         * The IEEE 802.11 standard this network is using. Will be one of the
+         * [ScanResult.WIFI_STANDARD_11N]..[ScanResult.WIFI_STANDARD_11BE] constants, or
+         * [ScanResult.WIFI_STANDARD_UNKNOWN].
+         */
+        val wifiStandard: Int,
     ) : WifiNetworkModel() {
         companion object {
             /**
@@ -255,11 +263,12 @@ sealed class WifiNetworkModel : Diffable<WifiNetworkModel> {
                 level: Int,
                 ssid: String? = null,
                 hotspotDeviceType: HotspotDeviceType = HotspotDeviceType.NONE,
+                wifiStandard: Int = ScanResult.WIFI_STANDARD_UNKNOWN,
             ): WifiNetworkModel {
                 if (!level.isValid()) {
                     return Inactive(getInvalidLevelErrorString(level))
                 }
-                return Active(showExclamation, level, ssid, hotspotDeviceType)
+                return Active(showExclamation, level, ssid, hotspotDeviceType, wifiStandard)
             }
 
             private fun Int.isValid(): Boolean {
@@ -304,6 +313,9 @@ sealed class WifiNetworkModel : Diffable<WifiNetworkModel> {
             if (prevVal.hotspotDeviceType != hotspotDeviceType) {
                 row.logChange(COL_HOTSPOT, hotspotDeviceType.name)
             }
+            if (prevVal.wifiStandard != wifiStandard) {
+                row.logChange(COL_WIFI_STANDARD, wifiStandard)
+            }
         }
 
         override fun logFull(row: TableRowLogger) {
@@ -314,6 +326,7 @@ sealed class WifiNetworkModel : Diffable<WifiNetworkModel> {
             row.logChange(COL_NUM_LEVELS, null)
             row.logChange(COL_SSID, ssid)
             row.logChange(COL_HOTSPOT, hotspotDeviceType.name)
+            row.logChange(COL_WIFI_STANDARD, wifiStandard)
         }
     }
 
@@ -368,6 +381,7 @@ const val COL_LEVEL = "level"
 const val COL_NUM_LEVELS = "maxLevel"
 const val COL_SSID = "ssid"
 const val COL_HOTSPOT = "hotspot"
+const val COL_WIFI_STANDARD = "wifiStandard"
 
 val LEVEL_DEFAULT: String? = null
 val NUM_LEVELS_DEFAULT: String? = null
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/binder/WifiViewBinder.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/binder/WifiViewBinder.kt
index 9f093a0fc..199d59b6f 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/binder/WifiViewBinder.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/binder/WifiViewBinder.kt
@@ -133,6 +133,14 @@ object WifiViewBinder {
                     }
                 }
 
+                launch {
+                    viewModel.wifiStandardIcon.collect { standardIcon ->
+                        wifiStandardView.isVisible = standardIcon != null
+                        if (standardIcon != null) {
+                            IconViewBinder.bind(standardIcon, wifiStandardView)
+                        }
+                    }
+                }
                 try {
                     awaitCancellation()
                 } finally {
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/compose/WifiIcon.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/compose/WifiIcon.kt
index f87323ee8..f1b4647d4 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/compose/WifiIcon.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/compose/WifiIcon.kt
@@ -19,10 +19,13 @@ package com.android.systemui.statusbar.pipeline.wifi.ui.compose
 import androidx.compose.foundation.layout.Box
 import androidx.compose.foundation.layout.Column
 import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.offset
 import androidx.compose.material3.LocalContentColor
 import androidx.compose.runtime.Composable
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.dp
+import com.android.systemui.common.shared.model.Icon
 import com.android.systemui.statusbar.pipeline.shared.ui.composable.ActivityIndicators
 import com.android.systemui.statusbar.shared.ui.compose.StatusBarIcon
 import com.android.systemui.statusbar.systemstatusicons.ui.viewmodel.SystemStatusIconViewModel
@@ -31,6 +34,7 @@ import com.android.systemui.statusbar.systemstatusicons.ui.viewmodel.SystemStatu
 @Composable
 fun WifiIcon(viewModel: SystemStatusIconViewModel.Wifi, modifier: Modifier = Modifier) {
     val icon = viewModel.icon ?: return
+    val standardIcon = viewModel.wifiStandardIcon
     Box(modifier = modifier) {
         if (viewModel.isActivityContainerVisible) {
             Row(verticalAlignment = Alignment.CenterVertically) {
@@ -41,10 +45,27 @@ fun WifiIcon(viewModel: SystemStatusIconViewModel.Wifi, modifier: Modifier = Mod
                         color = LocalContentColor.current,
                     )
                 }
-                StatusBarIcon(icon = icon)
+                WifiSignalWithStandard(icon = icon, standardIcon = standardIcon)
             }
         } else {
-            StatusBarIcon(icon = icon)
+            WifiSignalWithStandard(icon = icon, standardIcon = standardIcon)
+        }
+    }
+}
+
+@Composable
+private fun WifiSignalWithStandard(
+    icon: Icon,
+    standardIcon: Icon?,
+    modifier: Modifier = Modifier,
+) {
+    Box(modifier = modifier, contentAlignment = Alignment.Center) {
+        StatusBarIcon(icon = icon)
+        standardIcon?.let {
+            StatusBarIcon(
+                icon = it,
+                modifier = Modifier.offset(y = 1.5.dp),
+            )
         }
     }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/model/WifiIcon.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/model/WifiIcon.kt
index 8d53ee9f9..f05742eca 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/model/WifiIcon.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/model/WifiIcon.kt
@@ -129,13 +129,16 @@ sealed interface WifiIcon : Diffable<WifiIcon> {
 
         private fun WifiNetworkModel.Active.toBasicIcon(context: Context): Visible {
             val levelDesc = context.getString(WIFI_CONNECTION_STRENGTH[this.level])
-            return if (this.showExclamation) {
+            return if (!this.showExclamation) {
+                Visible(
+                    WifiIcons.WIFI_FULL_ICONS[this.level],
+                    ContentDescription.Loaded(levelDesc),
+                )
+            } else {
                 Visible(
                     WifiIcons.WIFI_NO_INTERNET_ICONS[this.level],
                     ContentDescription.Loaded("$levelDesc,${context.getString(NO_INTERNET)}"),
                 )
-            } else {
-                Visible(WifiIcons.WIFI_FULL_ICONS[this.level], ContentDescription.Loaded(levelDesc))
             }
         }
     }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiView.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiView.kt
index 0de5a8949..9b532b583 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiView.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiView.kt
@@ -84,6 +84,13 @@ class ModernStatusBarWifiView(context: Context, attrs: AttributeSet?) :
             lp.height =
                 resources.getDimensionPixelSize(R.dimen.status_bar_wifi_signal_height_updated)
 
+            val standardIconView = requireViewById<ImageView>(R.id.wifi_standard)
+            standardIconView.adjustViewBounds = true
+            val standardLp = standardIconView.layoutParams
+            standardLp.width = ViewGroup.LayoutParams.WRAP_CONTENT
+            standardLp.height =
+                resources.getDimensionPixelSize(R.dimen.status_bar_wifi_signal_height_updated)
+
             // New status bar icons have a single 3sp spacing defined
             (requireViewById<AlphaOptimizedLinearLayout>(R.id.wifi_group).layoutParams
                     as MarginLayoutParams)
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModel.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModel.kt
index acd307934..315180189 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModel.kt
@@ -17,10 +17,15 @@
 package com.android.systemui.statusbar.pipeline.wifi.ui.viewmodel
 
 import android.content.Context
+import android.net.wifi.ScanResult
+import com.android.systemui.common.shared.model.Icon
 import com.android.systemui.dagger.SysUISingleton
 import com.android.systemui.dagger.qualifiers.Background
 import com.android.systemui.log.table.TableLogBuffer
 import com.android.systemui.log.table.logDiffsForTable
+import com.android.systemui.res.R
+import com.android.systemui.shared.settings.data.repository.SecureSettingsRepository
+import com.android.systemui.statusbar.core.NewStatusBarIcons
 import com.android.systemui.statusbar.pipeline.dagger.WifiTableLog
 import com.android.systemui.statusbar.pipeline.shared.ConnectivityConstants
 import com.android.systemui.statusbar.pipeline.shared.data.model.DataActivityModel
@@ -35,6 +40,7 @@ import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.SharingStarted
 import kotlinx.coroutines.flow.StateFlow
 import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.distinctUntilChanged
 import kotlinx.coroutines.flow.flowOf
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.stateIn
@@ -56,6 +62,7 @@ constructor(
     interactor: WifiInteractor,
     @Background scope: CoroutineScope,
     wifiConstants: WifiConstants,
+    secureSettingsRepository: SecureSettingsRepository,
 ) : WifiViewModelCommon {
     override val wifiIcon: StateFlow<WifiIcon> =
         combine(
@@ -118,4 +125,45 @@ constructor(
                 activity.map { it != null && (it.hasActivityIn || it.hasActivityOut) }
             }
             .stateIn(scope, SharingStarted.WhileSubscribed(), false)
+
+    private val isWifiStandardDisplayEnabled: Flow<Boolean> =
+        secureSettingsRepository.boolSetting(
+            "wifi_standard",
+            context.resources.getBoolean(com.android.settingslib.R.bool.config_show_wifi_standard),
+        )
+
+    override val wifiStandardIcon: StateFlow<Icon?> =
+        combine(wifiIcon, interactor.wifiNetwork, isWifiStandardDisplayEnabled) {
+            wifiIcon,
+            network,
+            enabled,
+            ->
+            if (
+                !enabled ||
+                    !NewStatusBarIcons.isEnabled ||
+                    wifiIcon !is WifiIcon.Visible ||
+                    network !is WifiNetworkModel.Active
+            ) {
+                null
+            } else {
+                network.toWifiStandardIcon()
+            }
+        }
+            .distinctUntilChanged()
+            .stateIn(scope, SharingStarted.WhileSubscribed(), null)
+}
+
+private fun WifiNetworkModel.Active.toWifiStandardIcon(): Icon.Resource? {
+    if (this.showExclamation) {
+        return null
+    }
+    val resId =
+        when (this.wifiStandard) {
+            ScanResult.WIFI_STANDARD_11N -> R.drawable.ic_wifi_standard_4
+            ScanResult.WIFI_STANDARD_11AC -> R.drawable.ic_wifi_standard_5
+            ScanResult.WIFI_STANDARD_11AX -> R.drawable.ic_wifi_standard_6
+            ScanResult.WIFI_STANDARD_11BE -> R.drawable.ic_wifi_standard_7
+            else -> return null
+        }
+    return Icon.Resource(resId, null)
 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelCommon.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelCommon.kt
index feccbce75..044783671 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelCommon.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelCommon.kt
@@ -16,6 +16,7 @@
 
 package com.android.systemui.statusbar.pipeline.wifi.ui.viewmodel
 
+import com.android.systemui.common.shared.model.Icon
 import com.android.systemui.statusbar.pipeline.wifi.ui.model.WifiIcon
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.StateFlow
@@ -36,4 +37,7 @@ interface WifiViewModelCommon {
 
     /** True if the activity container view should be visible. */
     val isActivityContainerVisible: Flow<Boolean>
+
+    /** The Wi-Fi generation badge icon to be displayed over the Wi-Fi icon. */
+    val wifiStandardIcon: StateFlow<Icon?>
 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/ui/viewmodel/SystemStatusIconViewModel.kt b/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/ui/viewmodel/SystemStatusIconViewModel.kt
index 4d74a14ae..378bc82e4 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/ui/viewmodel/SystemStatusIconViewModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/ui/viewmodel/SystemStatusIconViewModel.kt
@@ -45,6 +45,7 @@ sealed interface SystemStatusIconViewModel {
          * a hydrated value.
          */
         val icon: Icon?
+        val wifiStandardIcon: Icon?
         /** True if we should always reserve space for the activity in/out indicators. */
         val isActivityContainerVisible: Boolean
         /** True if the activity in arrow should be visible. */
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/wifi/ui/viewmodel/WifiIconViewModel.kt b/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/wifi/ui/viewmodel/WifiIconViewModel.kt
index f7f020078..de127d31b 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/wifi/ui/viewmodel/WifiIconViewModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/systemstatusicons/wifi/ui/viewmodel/WifiIconViewModel.kt
@@ -50,6 +50,12 @@ constructor(@Assisted private val context: Context, wifiViewModel: WifiViewModel
             initialValue = WifiIcon.Hidden,
         )
 
+    override val wifiStandardIcon: Icon? by
+        wifiViewModel.wifiStandardIcon.hydratedStateOf(
+            traceName = "SystemStatus.wifiIcon.standardIcon",
+            initialValue = null,
+        )
+
     override val isActivityContainerVisible: Boolean by
         wifiViewModel.isActivityContainerVisible.hydratedStateOf(
             traceName = "SystemStatus.wifiIcon.activityContainerVisible",
diff --git a/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiViewTest.kt b/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiViewTest.kt
index bdd3daebe..4d193b85b 100644
--- a/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiViewTest.kt
+++ b/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/view/ModernStatusBarWifiViewTest.kt
@@ -30,11 +30,13 @@ import androidx.test.filters.SmallTest
 import com.android.systemui.SysuiTestCase
 import com.android.systemui.log.table.logcatTableLogBuffer
 import com.android.systemui.res.R
+import com.android.systemui.shared.settings.data.repository.fakeSecureSettingsRepository
 import com.android.systemui.statusbar.StatusBarIconView.STATE_DOT
 import com.android.systemui.statusbar.StatusBarIconView.STATE_HIDDEN
 import com.android.systemui.statusbar.StatusBarIconView.STATE_ICON
 import com.android.systemui.statusbar.core.NewStatusBarIcons
 import com.android.systemui.statusbar.phone.StatusBarLocation
+import com.android.systemui.statusbar.pipeline.ims.data.repository.CommonImsRepository
 import com.android.systemui.statusbar.pipeline.shared.ConnectivityConstants
 import com.android.systemui.statusbar.pipeline.shared.data.repository.FakeConnectivityRepository
 import com.android.systemui.statusbar.pipeline.shared.data.repository.connectivityRepository
@@ -48,6 +50,7 @@ import com.android.systemui.statusbar.pipeline.wifi.ui.viewmodel.LocationBasedWi
 import com.android.systemui.statusbar.pipeline.wifi.ui.viewmodel.LocationBasedWifiViewModel.Companion.viewModelForLocation
 import com.android.systemui.statusbar.pipeline.wifi.ui.viewmodel.WifiViewModel
 import com.android.systemui.testKosmos
+import org.mockito.kotlin.mock
 import com.google.common.truth.Truth.assertThat
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
@@ -85,7 +88,7 @@ class ModernStatusBarWifiViewTest : SysuiTestCase() {
         wifiRepository = FakeWifiRepository()
         wifiRepository.setIsWifiEnabled(true)
         scope = CoroutineScope(Dispatchers.Unconfined)
-        interactor = WifiInteractorImpl(connectivityRepository, wifiRepository, scope)
+        interactor = WifiInteractorImpl(connectivityRepository, wifiRepository, mock<CommonImsRepository>(), scope)
         val viewModelCommon =
             WifiViewModel(
                 connectivityConstants,
@@ -94,6 +97,7 @@ class ModernStatusBarWifiViewTest : SysuiTestCase() {
                 interactor,
                 scope,
                 wifiConstants,
+                kosmos.fakeSecureSettingsRepository,
             )
         viewModel = viewModelForLocation(viewModelCommon, StatusBarLocation.HOME)
     }
diff --git a/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelIconParameterizedTest.kt b/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelIconParameterizedTest.kt
index d8c1c25ab..454186532 100644
--- a/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelIconParameterizedTest.kt
+++ b/packages/SystemUI/tests/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelIconParameterizedTest.kt
@@ -24,11 +24,13 @@ import com.android.settingslib.AccessibilityContentDescriptions.WIFI_NO_CONNECTI
 import com.android.systemui.SysuiTestCase
 import com.android.systemui.common.shared.model.ContentDescription.Companion.loadContentDescription
 import com.android.systemui.log.table.TableLogBuffer
+import com.android.systemui.shared.settings.data.repository.fakeSecureSettingsRepository
 import com.android.systemui.statusbar.connectivity.WifiIcons.WIFI_FULL_ICONS
 import com.android.systemui.statusbar.connectivity.WifiIcons.WIFI_NO_INTERNET_ICONS
 import com.android.systemui.statusbar.connectivity.WifiIcons.WIFI_NO_NETWORK
 import com.android.systemui.statusbar.pipeline.airplane.data.repository.AirplaneModeRepository
 import com.android.systemui.statusbar.pipeline.airplane.data.repository.airplaneModeRepository
+import com.android.systemui.statusbar.pipeline.ims.data.repository.CommonImsRepository
 import com.android.systemui.statusbar.pipeline.shared.ConnectivityConstants
 import com.android.systemui.statusbar.pipeline.shared.data.model.ConnectivitySlot
 import com.android.systemui.statusbar.pipeline.shared.data.repository.FakeConnectivityRepository
@@ -58,6 +60,7 @@ import org.junit.runners.Parameterized.Parameters
 import org.mockito.Mock
 import org.mockito.Mockito.`when` as whenever
 import org.mockito.MockitoAnnotations
+import org.mockito.kotlin.mock
 
 @SmallTest
 @RunWith(Parameterized::class)
@@ -84,7 +87,7 @@ internal class WifiViewModelIconParameterizedTest(private val testCase: TestCase
         wifiRepository = FakeWifiRepository()
         wifiRepository.setIsWifiEnabled(true)
         scope = CoroutineScope(IMMEDIATE)
-        interactor = WifiInteractorImpl(connectivityRepository, wifiRepository, scope)
+        interactor = WifiInteractorImpl(connectivityRepository, wifiRepository, mock<CommonImsRepository>(), scope)
     }
 
     @After
@@ -116,6 +119,7 @@ internal class WifiViewModelIconParameterizedTest(private val testCase: TestCase
                     interactor,
                     scope,
                     wifiConstants,
+                    kosmos.fakeSecureSettingsRepository,
                 )
 
             val iconFlow = underTest.wifiIcon
diff --git a/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/domain/interactor/WifiInteractorKosmos.kt b/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/domain/interactor/WifiInteractorKosmos.kt
index 703619958..65662e9a8 100644
--- a/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/domain/interactor/WifiInteractorKosmos.kt
+++ b/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/domain/interactor/WifiInteractorKosmos.kt
@@ -18,14 +18,17 @@ package com.android.systemui.statusbar.pipeline.wifi.domain.interactor
 
 import com.android.systemui.kosmos.Kosmos
 import com.android.systemui.kosmos.applicationCoroutineScope
+import com.android.systemui.statusbar.pipeline.ims.data.repository.CommonImsRepository
 import com.android.systemui.statusbar.pipeline.shared.data.repository.connectivityRepository
 import com.android.systemui.statusbar.pipeline.wifi.data.repository.wifiRepository
+import org.mockito.kotlin.mock
 
 val Kosmos.wifiInteractor: WifiInteractor by
     Kosmos.Fixture {
         WifiInteractorImpl(
             connectivityRepository,
             wifiRepository,
+            mock<CommonImsRepository>(),
             applicationCoroutineScope,
         )
     }
diff --git a/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelKosmos.kt b/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelKosmos.kt
index 68c7fb2ce..2a8318f77 100644
--- a/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelKosmos.kt
+++ b/packages/SystemUI/tests/utils/src/com/android/systemui/statusbar/pipeline/wifi/ui/viewmodel/WifiViewModelKosmos.kt
@@ -20,6 +20,7 @@ import android.content.applicationContext
 import com.android.systemui.kosmos.Kosmos
 import com.android.systemui.kosmos.applicationCoroutineScope
 import com.android.systemui.log.table.logcatTableLogBuffer
+import com.android.systemui.shared.settings.data.repository.secureSettingsRepository
 import com.android.systemui.statusbar.pipeline.shared.ConnectivityConstants
 import com.android.systemui.statusbar.pipeline.wifi.domain.interactor.wifiInteractor
 import com.android.systemui.statusbar.pipeline.wifi.shared.WifiConstants
@@ -34,5 +35,6 @@ val Kosmos.wifiViewModel by
             wifiInteractor,
             applicationCoroutineScope,
             mock<WifiConstants>(),
+            secureSettingsRepository,
         )
     }
-- 
2.54.0

